homeassistant/components/srp_energy/config_flow.py

```diff
@@ -1,66 +1,106 @@
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
 
-from .const import CONF_IS_TOU, DEFAULT_NAME, SRP_ENERGY_DOMAIN
+from .const import CONF_IS_TOU, DEFAULT_NAME, DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
-
-class ConfigFlow(config_entries.ConfigFlow, domain=SRP_ENERGY_DOMAIN):
-    """Handle a config flow for SRP Energy."""
-
-    VERSION = 1
-
-    config = {
+DATA_SCHEMA = vol.Schema(
+    {
         vol.Required(CONF_ID): str,
         vol.Required(CONF_USERNAME): str,
         vol.Required(CONF_PASSWORD): str,
         vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
         vol.Optional(CONF_IS_TOU, default=False): bool,
     }
+)
 
-    async def async_step_user(self, user_input=None):
-        """Handle a flow initialized by the user."""
-        errors = {}
 
-        if self._async_current_entries():
-            return self.async_abort(reason="single_instance_allowed")
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
 
-        if user_input is not None:
-            try:
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    _LOGGER.debug("Validate User Input with client")
+    srp_client = SrpEnergyClient(
+        data[CONF_ID],
+        data[CONF_USERNAME],
+        data[CONF_PASSWORD],
+    )
 
-                srp_client = SrpEnergyClient(
-                    user_input[CONF_ID],
-                    user_input[CONF_USERNAME],
-                    user_input[CONF_PASSWORD],
-                )
+    is_valid = await hass.async_add_executor_job(srp_client.validate)
 
-                is_valid = await self.hass.async_add_executor_job(srp_client.validate)
+    _LOGGER.debug("Is valid User Input with client: %s", is_valid)
+    if not is_valid:
+        raise InvalidAuth
 
-                if is_valid:
-                    return self.async_create_entry(
-                        title=user_input[CONF_NAME], data=user_input
-                    )
+    return is_valid
 
-                errors["base"] = "invalid_auth"
 
-            except ValueError:
-                errors["base"] = "invalid_account"
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Unexpected exception")
-                errors["base"] = "unknown"
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle an SRP Energy config flow."""
+
+    VERSION = 1
 
+    @callback
+    def _show_form(self, errors: dict[str, Any] | None = None) -> FlowResult:
+        """Show the form to the user."""
+        _LOGGER.debug("Show Form")
         return self.async_show_form(
-            step_id="user", data_schema=vol.Schema(self.config), errors=errors
+            step_id="user",
+            data_schema=vol.Schema(
```

Comment:
Why does the schema on line 20 exist if we need to define it here (to access `self.hass`)?

Reply:
Good catch